packages/itwin/measure-tools/src/api/DrawingTypeDataCache.ts

```diff
@@ -3,42 +3,104 @@
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/
 
-import type { IModelConnection } from "@itwin/core-frontend";
+import { IModelConnection, Viewport } from "@itwin/core-frontend";
 import { IModelApp } from "@itwin/core-frontend";
 import { SheetMeasurementsHelper } from "./SheetMeasurementHelper";
+import { Id64String } from "@itwin/core-bentley";
 
-export class DrawingDataCache {
+class DrawingDataCache {
 
-  private _drawingTypeCache: SheetMeasurementsHelper.DrawingTypeData[];
-  private _sheetChangeListener: VoidFunction[] = [];
+  private _drawingTypeCache: Map<IModelConnection, Map<Id64String, SheetMeasurementsHelper.DrawingTypeData[]>>;
 
-  public get drawingtypes(): ReadonlyArray<SheetMeasurementsHelper.DrawingTypeData> {
-    return this._drawingTypeCache;
-  }
+  private _viewportModelChangedListeners: Map<Viewport, () => void>;
 
   public constructor() {
-    this._drawingTypeCache = [];
+    this._drawingTypeCache = new Map<IModelConnection, Map<Id64String, SheetMeasurementsHelper.DrawingTypeData[]>>();
+    this._viewportModelChangedListeners = new Map<Viewport, () => void>();
+
+    this.setupEvents();
+
+    // Populate initial viewports
+    for (const vp of IModelApp.viewManager)
+      this.addViewport(vp);
+  }
+
+  private setupEvents() {
+    // If an imodel closes, clear the cache for it
+    IModelConnection.onClose.addListener((imodel) => {
+      this._drawingTypeCache.delete(imodel);
+    });
+
+    // Listen for new viewports opening
+    IModelApp.viewManager.onViewOpen.addListener((vp) => {
+      this.addViewport(vp);
+    });
+
+    // Listen for viewports closing, this also is called when IModelApp shuts down
+    IModelApp.viewManager.onViewClose.addListener((vp) => {
+      this.dropViewport(vp);
+    });
   }
 
-  public async updateDrawingTypeCache(iModel: IModelConnection) {
-    this._sheetChangeListener.forEach((func) => {
-      func();
+  private addViewport(vp: Viewport) {
+    vp.onViewedModelsChanged.addListener((viewport) =>{
+      if (!viewport.view.isSheetView())
+        return;
+
+      void this.querySheetDrawingData(viewport.iModel, viewport.view.id);
     });
-    this._sheetChangeListener = [];
-    this._drawingTypeCache = [];
+  }
+
+  private dropViewport(vp: Viewport) {
+    const listener = this._viewportModelChangedListeners.get(vp);
+    if (listener) {
+      listener();
+      this._viewportModelChangedListeners.delete(vp);
+    }
+  }
+
+  public getSheetDrawingDataForViewport(vp: Viewport): SheetMeasurementsHelper.DrawingTypeData[] | [] {
```

Comment:
The return type should just be `ReadonlyArray<SheetMeasurementsHelper.DrawingTypeData>`,

Likewise for the query method, I think there's a difference between `Readonly<Array<T>>` and `ReadonlyArray<T>`.

Reply:
@Maxime-Brassard This was resolved without a code change, just have the return type `ReadonlyArray<SheetMeasurementsHelper.DrawingTypeData>`, best to convey that this is not a copy to callers so they don't mutate the internal data.

Reply:
@bsy-nicholasw Damn, thanks for spotting that, not sure how it hapenned... Maybe I missclosed it thinking it was another one.

Reply:
@Maxime-Brassard So I played with `Readonly<T[]>` and `ReadonlyArray<T>` in TS playground and it does seem like the type checker treats about as the same... however that empty `[]` you still have in the return type really doesn't make much sense to me.